modules/images/webp-uploads/load.php

```diff
@@ -573,3 +573,102 @@ function webp_uploads_update_featured_image( $html, $post_id, $attachment_id ) {
 	return webp_uploads_img_tag_update_mime_type( $html, 'post_thumbnail_html', $attachment_id );
 }
 add_filter( 'post_thumbnail_html', 'webp_uploads_update_featured_image', 10, 3 );
+
+/**
+ * Adds a fallback mechanism to replace webp images with jpeg alternatives on older browsers.
+ *
+ * @since n.e.x.t
+ */
+function webp_uploads_wepb_fallback() {
+	// Get mime type transofrms for the site.
+	$transforms = webp_uploads_get_upload_image_mime_transforms();
+	if ( ! is_array( $transforms ) ) {
+		return;
+	}
+
+	// We need to add fallback only if jpeg alternatives for the webp images are enabled for the server.
+	$preserve_jpegs_for_jpeg_transforms = in_array( 'image/jpeg', $transforms['image/jpeg'], true ) && in_array( 'image/webp', $transforms['image/jpeg'], true );
+	$preserve_jpegs_for_webp_transforms = in_array( 'image/jpeg', $transforms['image/webp'], true );
+	if ( ! $preserve_jpegs_for_jpeg_transforms && ! $preserve_jpegs_for_webp_transforms ) {
+		return;
+	}
+
+	?>
+	<script>
+		( function() {
+			window._fallbackWebpImages = function( media ) {
+				for ( var i = 0; i < media.length; i++ ) {
+					try {
+						var ext = media[i].media_details.sources['image/jpeg'].file.match( /\.\w+$/i );
+						if ( ! ext || ! ext[0] ) {
+							continue;
+						}
+
+						var images = document.querySelectorAll( 'img.wp-image-' + media[i].id );
+						for ( var j = 0; j < images.length; j++ ) {
+							images[j].src = images[j].src.replace( /\.webp$/i, ext[0] );
+							var srcset = images[j].getAttribute( 'srcset' );
+							if ( srcset ) {
+								images[j].setAttribute( 'srcset', srcset.replace( /\.webp(\s)/ig, ext[0] + '$1' ) );
+							}
+						}
+					} catch ( e ) {
+					}
+				}
+			};
+
+			var loadMediaDetails = function( nodes ) {
+				var ids = [];
+				for ( var i = 0; i < nodes.length; i++ ) {
+					if ( nodes[i].nodeName !== "IMG" || ! nodes[i].src.match( /\.webp$/i ) ) {
+						continue;
+					}
+
+					var attachment = nodes[i].className.match( /wp-image-(\d+)/i );
+					if ( attachment && attachment[1] && ids.indexOf( attachment[1] ) === -1 ) {
+						ids.push( attachment[1] );
+					}
+				}
+
+				for ( var page = 0, pages = Math.ceil( ids.length / 100 ); page < pages; page++ ) {
+					var pageIds = [];
+					for ( var i = 0; i < 100 && i + page * 100 < ids.length; i++ ) {
+						pageIds.push( ids[ i + page * 100 ] );
+					}
+
+					var jsonp = document.createElement( 'script' );
+					jsonp.src = '<?php echo esc_js( get_rest_url() ); ?>wp/v2/media/?_fields=id,media_details&_jsonp=_fallbackWebpImages&per_page=100&include=' + pageIds.join( ',' );
+					document.body.appendChild( jsonp );
+				}
+			};
```

Comment:
This is currently implementing approach 3 from https://github.com/WordPress/performance/issues/341#issuecomment-1143842022, which requires potentially a bunch of extra REST requests.

Given that all WebP images should always have JPEG available unless disabled on the entire setup (in which case this script isn't loaded anyway), I think the approach 2 from that implementation may be a more performant approach - just change file extensions from `.webp` to `.jpeg`. Otherwise, the REST requests may add a considerable amount of loading time, which would also result in the `img` tags not updated as quickly, thus causing the WebP file requests to not be cancelled.

The approach here is certainly a more reliable one - but it comes at the cost of performance I think, so we may still want to think about which of the two to go with further.

Reply:
> Given that all WebP images should always have JPEG available unless disabled on the entire setup (...), I think the approach 2 from that implementation may be a more performant approach - just change file extensions from .webp to .jpeg. 

@felixarntz, I went with the 2nd approach at the beginning but then it turned out that it wouldn't work well because JPEG images can have either `jpg` or `jpeg` extension and there is no way to know which one is used for an image without fetching this information from the server using REST APIs.

> This is currently implementing approach 3 from https://github.com/WordPress/performance/issues/341#issuecomment-1143842022, which requires potentially a bunch of extra REST requests.

Every request contains 100 images per page. I believe it's very unlikely to encounter a page that contains more than 100 images.

> Otherwise, the REST requests may add a considerable amount of loading time, which would also result in the img tags not updated as quickly, thus causing the WebP file requests to not be cancelled.

There is definitely a delay caused by the REST request and in some cases it causes slightly weird UI issues (images shows up as broken for a moment before a JPEG version is loaded). Unfortunately, it's a drawback of this approach, but I think that's fine because it will appear only for very old browser where such problem is not the worst one.

> The approach here is certainly a more reliable one - but it comes at the cost of performance I think, so we may still want to think about which of the two to go with further.

Unfortunately, JPEG extensions leave us only the third approach.

Reply:
@eugene-manuilov 
> > Given that all WebP images should always have JPEG available unless disabled on the entire setup (...), I think the approach 2 from that implementation may be a more performant approach - just change file extensions from .webp to .jpeg.
>
> @felixarntz, I went with the 2nd approach at the beginning but then it turned out that it wouldn't work well because JPEG images can have either `jpg` or `jpeg` extension and there is no way to know which one is used for an image without fetching this information from the server using REST APIs.

That's a fair point, thanks for clarifying. I didn't think about that, and I agree that the more reliable approach makes more sense here. Guessing the extension and then trying yet another time (e.g. first `jpg`, then `jpeg`) would be even more wasteful than the originally envisioned retry approach.